internal/transport/http2_server.go

```diff
@@ -1153,7 +1153,7 @@ func (t *http2Server) keepalive() {
 				if logger.V(logLevel) {
 					logger.Infof("transport: closing server transport due to maximum connection age.")
 				}
-				t.Close()
+				t.controlBuf.put(closeConnection{})
```

Comment:
I tried reading through the keepalive specs for the [client](https://github.com/grpc/proposal/blob/master/A8-client-side-keepalive.md) and [server](https://github.com/grpc/proposal/blob/master/A9-server-side-conn-mgt.md). There is no mention of implementations having to gracefully terminate the connection when the _keepalive timeout_ expires. The spec **does** talk about sending `GOAWAYs` in cases where the _idle timeout_ or _connection timeout + grace_ expire. So, I'm guessing we are doing the right thing here. But agree with you that the log statement can be improved.

Reply:
@ejona86 mentioned to me recently that we should always _try_ to disconnect cleanly by writing a GOAWAY.  I assume that applies here, too (and he may have mentioned this exact case, but I'm not fully remembering at this point).

Reply:
I think I was mentioning it in the context of abrupt failures after protocol errors and the like, because that's where we had the 1s timer in grpc java okhttp.

I think this case is more wishy-washy, because you _have_ sent GOAWAY well before this point. So I don't think HTTP/2 requires a third GOAWAY here. But you could do it for your own benefit. If the receiver sees GOAWAY with max_age_grace, it could be easier to debug what is going on. I think for Netty we just RST_STREAM the remaining streams, which is semi-clear that the closure is on purpose and hard to mistake for other things. We don't have GRACE behavior yet in the OkHttp server.

Reply:
> I think this case is more wishy-washy, because you have sent GOAWAY well before this point.

The case we are discussing now is the one where the client fails to respond to the server's keepalive ping, so there has not been a GOAWAY written at all yet.

Reply:
I'm adding logs for easier debugging throughout http2_client and http2_server. Changed that error message to ""transport: closing server transport due to keep alive ping not being acked within timeout %v.", t.kp.Time.String()", will be out in a future PR.

Reply:
See https://github.com/grpc/grpc-go/pull/5840/files#diff-4058722211b8d52e2d5b0c0b7542059ed447a04017b69520d767e94a9493409eR1179.